controllers/metal3.io/bmceventsubscription_controller.go

```diff
@@ -0,0 +1,260 @@
+	}
+
+	httpHeaders, err := r.getHTTPHeaders(*subscription)
+
+	if err != nil {
+		reqLogger.Error(err, "failed to get http headers")
+		return ctrl.Result{RequeueAfter: subscriptionRetryDelay}, nil
+	}
+
+	if subscription.DeletionTimestamp.IsZero() {
+		// Not being deleted
+		if subscription.Status.SubscriptionID == "" {
+			if _, err := prov.AddBMCEventSubscriptionForNode(subscription, httpHeaders); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to add a subscription")
+			}
+
+			// Update the subscription with the new ID
+			err := r.Status().Update(ctx, subscription)
+			if err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to update subscription status")
+			}
+
+		} else {
+			// Update an existing subscription by recreating it in the BMC
+
+			originalSubscription := subscription.DeepCopy()
+
+			if _, err := prov.AddBMCEventSubscriptionForNode(subscription, httpHeaders); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to update subscription")
+			}
+
+			if _, err := prov.RemoveBMCEventSubscriptionForNode(*originalSubscription); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to remove a subscription")
+			}
+
+			return ctrl.Result{}, nil
+		}
+	} else {
+		// being deleted
+		if subscriptionHasFinalizer(subscription) {
+			if _, err := prov.RemoveBMCEventSubscriptionForNode(*subscription); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to remove a subscription")
+			}
+
+			// Remove finalizer to allow deletion
+			subscription.Finalizers = utils.FilterStringFromList(
+				subscription.Finalizers, metal3v1alpha1.BMCEventSubscriptionFinalizer)
+			reqLogger.Info("cleanup is complete, removed finalizer",
+				"remaining", subscription.Finalizers)
+			if err := r.Update(context.Background(), subscription); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+
+		// Stop reconciliation as the item is being deleted
+		return ctrl.Result{}, nil
+	}
+
+	return
+
+}
+
+func (r *BMCEventSubscriptionReconciler) getBMCSecret(request ctrl.Request, host *metal3v1alpha1.BareMetalHost) (bmcCredsSecret *corev1.Secret, err error) {
+	secretKey := host.CredentialsKey()
+	bmcCredsSecret = &corev1.Secret{}
+	err = r.Get(context.TODO(), secretKey, bmcCredsSecret)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil, &ResolveBMCSecretRefError{message: fmt.Sprintf("The BMC secret %s does not exist", secretKey)}
+		}
+		return nil, err
+	}
+
+	return bmcCredsSecret, nil
+}
+
+func (r *BMCEventSubscriptionReconciler) getHTTPHeaders(subscription metal3v1alpha1.BMCEventSubscription) ([]map[string]string, error) {
+	headers := []map[string]string{}
+
+	if subscription.Spec.HTTPHeadersRef == nil {
+		return headers, nil
+	}
+
+	secret := &corev1.Secret{}
+	secretKey := types.NamespacedName{
+		Name:      subscription.Spec.HTTPHeadersRef.Name,
+		Namespace: subscription.Spec.HTTPHeadersRef.Namespace,
+	}
+
+	err := r.Get(context.TODO(), secretKey, secret)
```

Comment:
Note: I'm not sure if it could have any impact here, but the main manager has a cache selector to avoid caching all the secrets, so I guess also such secret should have the environment label (in BMO there's a `SecretManager` [class](https://github.com/metal3-io/baremetal-operator/blob/c22cb7ba6df61e7a4b3bae751f3b36b0ae825f1c/pkg/secretutils/secret_manager.go#L34) to keep care of it, probably worth checking)